Refactor convert_IN_to_antijoin.

kainwen opened the pull request:
Greenplum pulls up NOT-IN sublink to achieve better OLAP performance.
Generally, NOT-IN sublink is not semantically equivalent to NOT-EXIST
sublink because of NULL values. Besides, Greenplum cannot choose Hash
Join and cannot take advantage of partition-distribution  due to the
limitation of Executor currently (refer f77bf0879a28 and the comments
https://github.com/greenplum-db/gpdb/pull/11579#issuecomment-793385278).
Life will be much easier if both sides of NOT-IN sublink testexpr are
sure non-nullable.

We might seek for a great improvement of the executor to better support
NOT_IN. This commit just refactors the code and make the non-nullable
deduction clearer and can cover more cases. It does the following things:

  * Rename the function find_nonnullable_vars_walker.
    I find that cdbsubselect.c:find_nonnullable_vars_walker shares the
    same name with clauses.c:find_nonnullable_vars_walker even both
    are qualified by static.

  * Remove is_targetlist_nullable and is_param_nullable.
    These two functions almost do the same thing (testing a list of
    exprs if non-nullable).

  * Introduce new logic to test non-nullable.
    1. build a list of exprs from the source (either the inner subselect's
       targetlist or the outer part of the sublink's testexpr)
    2. build a set of non-nullable vars
    3. try our best to deduct non-nullable of the list of exprs.

See github issue: https://github.com/greenplum-db/gpdb/issues/12930

Co-authored-by: Tao Tang <tang.tao.cn@gmail.com>

khannaekta commented on src/backend/cdb/cdbsubselect.c:
```diff
@@ -1178,182 +1181,195 @@ find_nonnullable_vars_walker(Node *node, NonNullableVarsContext *context)
+		 * in other cases.
+		 */
+		if (IsA(expr, OpExpr))
+			return list_make1(linitial(((OpExpr *)expr)->args));
```
This assumption is true for single-column `not-in` but in case of a multi-column `<>ALL` we should traverse through the entire list of `expr->args`.

for the test example with NULLable on the sublink, 
```
truncate t1_12930;
truncate t2_12930;
alter table t2_12930 alter column b set not null;
alter table t1_12930 alter column b drop not null;
insert into t1_12930 values (1, null);
insert into t2_12930 values (1, 1);
select * from t1_12930 where (a, b) <>ALL (select a, b from t2_12930);
```

-- expected 
```
select * from t1_12930 where (a, b) <>ALL (select a, b from t2_12930) ;
 a | b
---+---
(0 rows)
```

-- current results
```
select * from t1_12930 where (a, b) <>ALL (select a, b from t2_12930) ;
 a | b
---+---
 1 |
(1 row)
```

kainwen replied:
> truncate t1_12930;
> truncate t2_12930;
> alter table t2_12930 alter column b set not null;
> alter table t1_12930 alter column b drop not null;
> insert into t1_12930 values (1, null);
> insert into t2_12930 values (1, 1);
> select * from t1_12930 where (a, b) <>ALL (select a, b from t2_12930);

Ah, good catch @khannaekta 
Thanks so much! Let refactor the code and add more case.

khannaekta reviewed (commented):
Overall, the logic looks fine, but can we double check `fetch_outer_exprs()` to be sure we do not miss any case, just as the one we see with `<>ALL`.

kainwen commented:
> Overall, the logic looks fine, but can we double check `fetch_outer_exprs()` to be sure we do not miss any case, just as the one we see with `<>ALL`.

@khannaekta Thanks for you comments!

I have just refactored the code and added a test case (based on yours).

chrishajas commented:
Overall LGTM (though admittedly don't have much context in this area of code), I'll do more testing for it tomorrow.

khannaekta reviewed (commented):
LGTM!

kainwen commented:
Pipeline fails due to known flaky case. Dev pipeline is Green. Push this.

Thank all for the comments!
